### ISSUE TITLE
Reserving nicknames

### ISSUE THREAD

**Issue by DanielOaks** [oragono/oragono#169]
This is something that might be desirable to certain networks. It's also something that really... makes implementation strange. There's a lot to think about around this issue wrt approaches and such.

I think there's one decent way to handle this. One bool in the config file that turns on/off nickname registration + reservation.

If the bool is on, each account name acts as a reserved nickname. There's a new key on accounts, `grouped-nicks` (managed through traditional `GROUP`/etc commands with our fake NickServ client). You can `NICKSERV GHOST` people using your nickname, and people who change to a nick but aren't authenticated are asked to login, but not renamed off automagically unless the user's specifically asked for that (another bool, something like `kill-nick-squatters` that defaults to false and can be enabled with a call to NickServ).

If the bool is off, no reserving of nicknames exists at all. In addition, any existing `grouped-nicks` information is left alone, not updated nor removed. Any `GROUP`/related commands on NickServ fail out instantly with the reply _"Nickname registration is not enabled"_.

We'll want to show appropriate help information for `NICKSERV REGISTER` whether nickname reservation is enabled or disabled. That's gonna be an interesting thing to look at.

**Comment by slingamn** [comment]
I definitely agree that it should be possible to disable nickname registration. I would be interested in more modes besides "disabled" and "lenient enforcement":

1. Full enforcement: if a nickname is registered, attempts to change to it with `/nick` fail unless the user is logged in. (This would match a configuration where everyone is expected to log in with SASL.)
2. Freenode-style enforcement with a timeout: an unauthenticated user who sets a reserved nick has 30 seconds (configurable) to either authenticate or switch, otherwise they're disconnected (or their nick is set to something random that starts with `guest`).

I don't think these would add much implementation complexity. Any thoughts on whether they would interact negatively with current conventions/standards?

**Comment by DanielOaks** [comment]
I could see that sorta thing. Nick registration config looking something like:
```yaml
nick-registration:
    enabled: true
    enforcement: none / timeout / disable-nick-entirely
```

The `none` and `timeout` options are already out there and in use today, so no trouble at all.

Stopping users from changing their nick to it at all isn't out there today, buuuut... so long as you output an appropriate message it'd work. Honestly, the main concern with the disabling the nick entirely is how to manage logging users in (using `NICKSERV`) while their nickname isn't the one they're trying to login as and similar.

**Comment by slingamn** [comment]
Yeah, you'd have to use SASL.

**Comment by DanielOaks** [comment]
I should note, there's a doc that discusses this and the different ways networks do it today: https://gist.github.com/grawity/6733799

**Comment by slingamn** [comment]
Status: this is partially implemented. Here are the remaining features:

- [ ] support for `NICKSERV GHOST`
- [ ] support for reserving multiple nicknames per account, up to a limit
- [ ] the ability to specify a custom enforcement mode for one's own account (so even if the server defaults to `timeout` enforcement, you can set your account to `strict`); this can be disabled at runtime in the config
- [ ] a `lenient` enforcement mode, where there is no enforcement by default, but we respect people's per-account opt-ins

**Comment by slingamn** [comment]
This was fully implemented in #206 and #324!
